### ISSUE TITLE
unify_with_occurs_check/2 creates infinite tree, dumping all cores

### ISSUE THREAD

**Issue by UWN** [trealla-prolog/trealla#532]
```
Trealla Prolog (c) Infradig 2020-2022, v1.26.15-1-ga85089
?- \B=A,C=[B|A].
   A = \B, C = [B|\B]. % expected
?- \B=A,C=[B|A],unify_with_occurs_check(B,[C|A]).
   B = [[[...]|...]]|\...]|...]]]|\[...]|...]]|...]]]|\[[...]|...]]|...]]|\...]|...]]]], A = \B, C = [[[...]|...]]|\...]|...]]]|\[...]|...]]|...]]]|\[[...]|...]]|...]]|\...]|...]]]]. % unexpected, dumping core (13211-1, 8.2.2.1 and 7.3)
?- \B=A,C=[B|A],unify_with_occurs_check(B,[C|A]), acyclic_term(B).
   false. % unexpected, as above
?- \B=A,C=[B|A],unify_with_occurs_check(B,[C|A]), unify_with_occurs_check(A,\C).
Segmentation fault (core dumped) % explicit message about dumping core
```

Those who dump core, dump core.

**Comment by infradig** [comment]
Strange, using an intermediate variable it works...

?- \B=A,C=[B|A],unify_with_occurs_check(B,[C|A]).

   B = [[[...]|...]]|\...]|...]]]|\[...]|...]]|...]]]|\[[...]|...]]|...]]|\...]|...]]]],
A = \B, C = [[[...]|...]]|\...]|...]]]|\[...]|...]]|...]]]|\[[...]|...]]|...]]|\...]|...]]]].


?- \B=A, A=D, C=[B|D], unify_with_occurs_check(B,[C|A]).
   false.

**Comment by UWN** [comment]
Can you put into simple word what the source of this error was, such that I can adjust my testing accordingly?

**Comment by infradig** [comment]
It was going down the rabbit hole leaving crumbs, but forgetting to pick
them up on the way out. When going in again and laying a new trail it got
confused coming out as it encountered crumbs and didn't know if it had been
here before.

**Comment by UWN** [comment]
So in a sense there was a built-in that left some inconsistent state. Right?

(Recall that the [Dedekind numbers](https://en.wikipedia.org/wiki/Dedekind_number) get soon very large and thus I cannot find all errors by brute force alone as I do now)

**Comment by infradig** [comment]
Yes. I have become lax about adding tests for issues, which has let
problems reappear. I have added these ones.

**Comment by UWN** [comment]
If you think some issue is a reappearance, please indicate. My mental model of actual Prolog processors is often only half, no rather only a quadrant of what they are, but then I am not an implementer.

Speaking of quad-rants, soon questions using answer descriptions (quads) will simplify regression testing substantially. (I for the moment do not do any regression testing. I only dive into the deep, as it is much easier for me...)

**Comment by pmoura** [comment]
> Yes. I have become lax about adding tests for issues, which has let problems reappear. I have added these ones.

Maybe also consider **contributing** those tests to the [automated tests suite](https://trealla-prolog.github.io/trealla/) you're already using? So that others can also benefit from those new tests?

**Comment by infradig** [comment]
touché

**Comment by UWN** [comment]
> touché

Better put your time into improving your system. There is very little value in compiling them (manually) into some extra form. Yes, there is some value, but very little indeed. Not worth the typing (cut-paste) time. Instead, removing current issues enables me to find further errors. Currently, everything is stalled again. And due to your systematic under-appreciation of syntax, you are just increasing the MTTFF.

**Comment by pmoura** [comment]
> > touché
> 
> Better put your time into improving your system. There is very little value in compiling them (manually) into some extra form. Yes, there is some value, but very little indeed.

Automated testing have uncovered ~1000 standards compliance bugs so far in Trealla. That greatly outnumbers the bugs 
found using manual testing. But not everything can be (easily) automated and both testing approaches contribute to improve the quality of Prolog systems. Tests can only prove the existence of bugs, not their absence. But automating tests greatly helps prevent regressions, specially when bug fixes are committed together with tests that exposed the bugs.

**Comment by UWN** [comment]
Nobody hinders you to use the witnesses posted here to improve your test suite. Fine. But as far as I know this test suite is on a password protected github site.

But I can assure you that the precise errors we currently encounter w.r.t `(=)/2` and `dif/2` are not very worthy being recorded. The administrative overheads are much too high w.r.t. the benefits.

**Comment by pmoura** [comment]
> But as far as I know this test suite is on a password protected github site.

Wrong. The test suite is open source. It was never on a password protected site. You can access it at:

https://github.com/LogtalkDotOrg/logtalk3

It's also part and therefore included in all Logtalk installations, which can be easily performed using publicly avaialble installers.
